nemo_aligner/utils/text_generation_utils.py

```diff
@@ -14,11 +14,60 @@
 
 """Utilities for generating text."""
 
+from typing import Any, List
 
 import torch
+
+from megatron.core import parallel_state
+from nemo.collections.nlp.modules.common.text_generation_strategy import GPTModelTextGenerationStrategy
 from nemo.utils import logging
 
 
+class TrackLengthGPTModelTextGenerationStrategy(GPTModelTextGenerationStrategy):
+    """
+    Text generation strategy that tracks the length of the generated text.
+
+    TODO This is a temporary workaround until NeMo's `generate()` function returns this information.
+    """
+
+    def __init__(self, model: Any, context_lengths: torch.Tensor, max_length: int):
+        super().__init__(model)
+        self._context_lengths = context_lengths
+        self._max_length = max_length
+        self._end_idx = torch.full_like(context_lengths, fill_value=-1)
+
+    def end_of_generation_condition(
+        self, tokens: torch.Tensor, prev: torch.Tensor, eod_id: int, end_strings: List[str]
+    ) -> torch.Tensor:
+        is_end = super().end_of_generation_condition(tokens=tokens, prev=prev, eod_id=eod_id, end_strings=end_strings)
+        assert len(is_end) == len(tokens)
+        if len(tokens) != len(self._context_lengths):
+            raise RuntimeError(
+                "Batch size mismatch: the `context_lengths` tensor provided in the constructor has batch size "
+                f"{len(self._context_lengths)}, while the generated tokens have batch size {len(tokens)}"
+            )
```

Comment:
this logic looks to guard against the batch size changing, in what cases can this occur?

Reply:
In case someone instantiate a `TrackLengthGPTModelTextGenerationStrategy` with an incorrect `context_lengths`. Definitely not supposed to happen with the current code but since this is a standalone class (and I wanted to make sure my code was correct) I preferred to add a check.